src/mlpack/methods/ann/layer/bilinear_interpolation_impl.hpp

```diff
@@ -0,0 +1,173 @@
+ * 3-clause BSD license along with mlpack. If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_METHODS_ANN_LAYER_BILINEAR_INTERPOLATION_IMPL_HPP
+#define MLPACK_METHODS_ANN_LAYER_BILINEAR_INTERPOLATION_IMPL_HPP
+
+// In case it hasn't yet been included.
+#include "bilinear_interpolation.hpp"
+
+namespace mlpack {
+namespace ann /** Artificial Neural Network. */ {
+
+
+template<typename InputDataType, typename OutputDataType>
+BilinearInterpolation<InputDataType, OutputDataType>::
+BilinearInterpolation():
+    inRowSize(0),
+    inColSize(0),
+    outRowSize(0),
+    outColSize(0),
+    depth(0)
+{
+  // Nothing to do here.
+}
+
+template<typename InputDataType, typename OutputDataType>
+BilinearInterpolation<InputDataType, OutputDataType>::
+BilinearInterpolation(
+    const size_t inRowSize,
+    const size_t inColSize,
+    const size_t outRowSize,
+    const size_t outColSize,
+    const size_t depth):
+    inRowSize(inRowSize),
+    inColSize(inColSize),
+    outRowSize(outRowSize),
+    outColSize(outColSize),
+    depth(depth)
+{
+  // Nothing to do here.
+}
+
+template<typename InputDataType, typename OutputDataType>
+template<typename eT>
+void BilinearInterpolation<InputDataType, OutputDataType>::Forward(
+    const arma::Mat<eT>&& input, arma::Mat<eT>&& output)
+{
+  if (output.is_empty())
+    output.set_size(outRowSize * outColSize * depth, 1);
+
+  assert(input.n_rows == inRowSize * inColSize * depth);
+  assert(input.n_cols == 1);
+  assert(output.n_rows == outRowSize * outColSize * depth);
+  assert(output.n_cols == 1);
+  assert(inRowSize >= 2);
+  assert(inColSize >= 2);
+
+  arma::cube inputAsCube(input.memptr(), inRowSize, inColSize, depth);
+  arma::cube outputAsCube(output.memptr(), outRowSize, outColSize, depth, false, true);
+
+  double scaleRow = (double) inRowSize / (double) outRowSize;
+  double scaleCol = (double) inColSize / (double) outColSize;
+
+  arma::mat22 coeffs;
+  for (size_t i = 0; i < outRowSize; i++)
+  {
+    size_t rOrigin = (size_t) std::floor(i * scaleRow);
+    if (rOrigin > inRowSize - 2)
+      rOrigin = inRowSize - 2;
+    double deltaR = i * scaleRow - rOrigin;
+    for (size_t j = 0; j < outColSize; j++)
+    {
+      size_t cOrigin = (size_t) std::floor(j * scaleCol);
+
+      if (cOrigin > inColSize - 2)
+        cOrigin = inColSize - 2;
+
+      double deltaC = j * scaleCol - cOrigin;
+      coeffs[0] = (1 - deltaR) * (1 - deltaC);
+      coeffs[1] = deltaR * (1 - deltaC);
+      coeffs[2] = (1 - deltaR) * deltaC;
+      coeffs[3] = deltaR * deltaC;
+
+      for (size_t k = 0; k < depth; k++)
+      {
+        outputAsCube(i, j, k) = arma::accu(inputAsCube.slice(k).submat(
+                rOrigin, cOrigin, rOrigin + 1, cOrigin + 1) % coeffs);
+      }
+    }
+  }
+}
+
+template<typename InputDataType, typename OutputDataType>
+template<typename eT>
+void BilinearInterpolation<InputDataType, OutputDataType>::Backward(
```

Comment:
Do you think it make sense, to write a function for the interpolation and just pass the correct parameter, since the forward and backward pass is almost the same?

Reply:
It is doable in a function, but we'd have to specify whether we want a forward or a backward pass, and introduce a number of `if-else` statements. Also, a number of variables have the same name, but are not computed the same way in both the function (see `scaleRow` and `scaleCol` in both the functions). I personally prefer the current implementation, but I'd say its your call.

Reply:
Okay, let's go with the current implementation.